includes/class-signature.php

```diff
@@ -107,12 +124,244 @@ public static function generate_signature( $user_id, $http_method, $url, $date,
+		if ( is_object( $request ) ) { // REST Request object
+			$headers = $request->get_headers();
+			$headers['(request-target)'][0] = strtolower( $request->get_method() ) . ' /' . rest_get_url_prefix() . $request->get_route();
+		} else {
+			$request = self::format_server_request( $request );
+			$headers = $request['headers']; // $_SERVER array
+			$headers['(request-target)'][0] = strtolower( $headers['request_method'][0] ) . ' ' . $headers['request_uri'][0];
+		}
+
+		if ( ! isset( $headers['signature'] ) ) {
+			return new \WP_Error( 'activitypub_signature', 'Request not signed', array( 'status' => 403 ) );
+		}
+
+		if ( array_key_exists( 'signature', $headers ) ) {
+			$signature_block = self::parse_signature_header( $headers['signature'] );
+		} elseif ( array_key_exists( 'authorization', $headers ) ) {
+			$signature_block = self::parse_signature_header( $headers['authorization'] );
+		}
+
+		if ( ! isset( $signature_block ) || ! $signature_block ) {
+			return new \WP_Error( 'activitypub_signature', 'Incompatible request signature. keyId and signature are required', array( 'status' => 403 ) );
+		}
+
+		$signed_headers = $signature_block['headers'];
+		if ( ! $signed_headers ) {
+			$signed_headers = array( 'date' );
+		}
+
+		$signed_data = self::get_signed_data( $signed_headers, $signature_block, $headers );
+		if ( ! $signed_data ) {
+			return new \WP_Error( 'activitypub_signature', 'Signed request date outside acceptable time window', array( 'status' => 403 ) );
+		}
+
+		$algorithm = self::get_signature_algorithm( $signature_block );
+		if ( ! $algorithm ) {
+			return new \WP_Error( 'activitypub_signature', 'Unsupported signature algorithm (only rsa-sha256 and hs2019 are supported)', array( 'status' => 403 ) );
+		}
+
+		if ( \in_array( 'digest', $signed_headers, true ) && isset( $body ) ) {
+			if ( is_array( $headers['digest'] ) ) {
+				$headers['digest'] = $headers['digest'][0];
+			}
+			$digest = explode( '=', $headers['digest'], 2 );
+			if ( 'SHA-256' === $digest[0] ) {
+				$hashalg = 'sha256';
+			}
+			if ( 'SHA-512' === $digest[0] ) {
+				$hashalg = 'sha512';
+			}
+
+			if ( \base64_encode( \hash( $hashalg, $body, true ) ) !== $digest[1] ) { // phpcs:ignore
+				return new \WP_Error( 'activitypub_signature', 'Invalid Digest header', array( 'status' => 403 ) );
+			}
+		}
+
+		$public_key = self::get_remote_key( $signature_block['keyId'] );
+		if ( \is_wp_error( $public_key ) ) {
+			return $public_key;
+		}
+		$verified = \openssl_verify( $signed_data, $signature_block['signature'], $public_key, $algorithm ) > 0;
+		if ( ! $verified ) {
+			return new \WP_Error( 'activitypub_signature', 'Invalid signature', array( 'status' => 403 ) );
+		}
+		return $verified;
+	}
+
+	/**
+	 * Get public key from key_id
+	 *
+	 * @param string $key_id
+	 * @return string $publicKeyPem
+	 * @author Django Doucet <django.doucet@webdevstudios.com>
+	 */
+	public static function get_remote_key( $key_id ) { // phpcs:ignore
```

Comment:
don't we have an actor URL here? can we always be sure, that the key_id is the actur URL?

Reply:
I think we have the request here, why not use the Actor-URL instead?

Reply:
I think the content-negotiated get requests didn't include an actor.

I haven't yet seen an implementation whose keyId wasn't added to the actor url:
- #main-key
- /main-key

Reply:
Can you make post an example how the request looks??

Reply:
Here are some example requests from Mastodon

```
WP_REST_Request Object
(
    [method:protected] => POST
    [params:protected] => Array
        (

            [JSON] => Array
                (
                    [@context] => Array
                        (
                            [0] => https://www.w3.org/ns/activitystreams
                            [1] => Array
                                (
                                    [ostatus] => http://ostatus.org#
                                    [atomUri] => ostatus:atomUri
                                    [inReplyToAtomUri] => ostatus:inReplyToAtomUri
                                    [conversation] => ostatus:conversation
                                    [sensitive] => as:sensitive
                                    [toot] => http://joinmastodon.org/ns#
                                    [votersCount] => toot:votersCount
                                )

                        )

                    [id] => https://masto.example.app/users/mastodon/statuses/110082446069839861/activity
                    [type] => Create
                    [actor] => https://masto.example.app/users/mastodon
                    [published] => 2023-03-25T06:13:44Z
                    [to] => Array
                        (
                            [0] => https://www.w3.org/ns/activitystreams#Public
                        )

                    [cc] => Array
                        (
                            [0] => https://masto.example.app/users/mastodon/followers
                            [1] => https://wp-example.ngrok.app/author/wordpress/
                        )

                    [object] => Array
                        (
                            [id] => https://masto.example.app/users/mastodon/statuses/110082446069839861
                            [type] => Note
                            [summary] => 
                            [inReplyTo] => https://wp-example.ngrok.app/sad-eros/
                            [published] => 2023-03-25T06:13:44Z
                            [url] => https://masto.example.app/@mastodon/110082446069839861
                            [attributedTo] => https://masto.example.app/users/mastodon
                            [to] => Array
                                (
                                    [0] => https://www.w3.org/ns/activitystreams#Public
                                )

                            [cc] => Array
                                (
                                    [0] => https://masto.example.app/users/mastodon/followers
                                    [1] => https://wp-example.ngrok.app/author/wordpress/
                                )

                            [sensitive] => 
                            [atomUri] => https://masto.example.app/users/mastodon/statuses/110082446069839861
                            [inReplyToAtomUri] => https://wp-example.ngrok.app/sad-eros/
                            [conversation] => tag:masto.example.app,2023-03-18:objectId=35:objectType=Conversation
                            [content] => <p><span class="h-card"><a href="https://wp-example.ngrok.app/author/wordpress/" class="u-url mention">@<span>wordpress</span></a></span> $request</p>
                            [contentMap] => Array
                                (
                                    [en] => <p><span class="h-card"><a href="https://wp-example.ngrok.app/author/wordpress/" class="u-url mention">@<span>wordpress</span></a></span> $request</p>
                                )

                            [attachment] => Array
                                (
                                )

                            [tag] => Array
                                (
                                    [0] => Array
                                        (
                                            [type] => Mention
                                            [href] => https://wp-example.ngrok.app/author/wordpress/
                                            [name] => @wordpress@wp-example.ngrok.app
                                        )

                                )

                            [replies] => Array
                                (
                                    [id] => https://masto.example.app/users/mastodon/statuses/110082446069839861/replies
                                    [type] => Collection
                                    [first] => Array
                                        (
                                            [type] => CollectionPage
                                            [next] => https://masto.example.app/users/mastodon/statuses/110082446069839861/replies?only_other_accounts=true&page=true
                                            [partOf] => https://masto.example.app/users/mastodon/statuses/110082446069839861/replies
                                            [items] => Array
                                                (
                                                )

                                        )

                                )

                        )

                    [signature] => Array
                        (
                            [type] => RsaSignature2017
                            [creator] => https://masto.example.app/users/mastodon#main-key
                            [created] => 2023-03-25T06:13:44Z
                            [signatureValue] => Ic9JGB+Q3DkYnx/LlGmOkW6eYJbfhqaQxaDEeL74urpOpKYwjFOfPSyHZkIiGW5kzqaAqC8OSp3Yzds441yoQZSLruWEpjL764o9J9FKkYM4cZr7lBasq8tdRtYLag0hEJtgt9yvXS7JiM3Tk4F/5XVrQ1B9D4OtQQwBtILB1zUifzocqi+x31VeV45fwkRI61yjFhhOE7NTHB0P9hfnrArP9OcCKYqISqOgH2qApGXVeZgs4hW5LMxjJyXG+KLBmyNFiaLBCyyAgiwKxgDzeOGMwbisEnSxCZwyLTuEeqAp7HUbj3FVQZUlbLVN/B678Q2jVakyN+mAe6sUHsKV1w==
                        )

                )

            [defaults] => Array
                (
                )

        )

    [headers:protected] => Array
        (
            [x_original_host] => Array
                (
                    [0] => wp-example.ngrok.app
                )

            [signature] => Array
                (
                    [0] => keyId="https://masto.example.app/users/mastodon#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest content-type",signature="Cva/xbz0ntRh6qsrFANPkiXGMTdfWy03+uK7Hqb7QQW/QpgUZkK1A6iQLtcpiI1rTIStyhFc96ySdwtzVHjLRr40dRqQAaqyxJ8BkXh7ScNwS2lVv4j6LXQJSNUNkY/Vcs+na+4++Lc80Pl82AFQPODAxrOuwmbdmOkqCFNu/lQ2I1JvG3WgsTgQrntlruS7oYZQ/9UyUnG3nWi3mYrqnKtgmPc5MFLkDFSsreHFD9KCW932fvn5VCpmaFxKS+PVmXbRVgGBh7UZHS+hTRgxrmwAL8VXc5V0QqHEm5zb5JpjvcLJ+OjZ8l0T1J73uHvis4sZY9sExYPSHDSRKc27uQ=="
                )

            [digest] => Array
                (
                    [0] => SHA-256=hdi8H0pMg8q8KebwR8/HIEZFRBW5V6tNprKvxAFm8xE=
                )

            [date] => Array
                (
                    [0] => Sat, 25 Mar 2023 06:13:44 GMT
                )

            [content_type] => Array
                (
                    [0] => application/activity+json
                )

            [accept_encoding] => Array
                (
                    [0] => gzip
                )

            [user_agent] => Array
                (
                    [0] => http.rb/5.1.0 (Mastodon/4.0.2; +https://masto.example.app/)
                )

            [content_length] => Array
                (
                    [0] => 3035
                )

            [host] => Array
                (
                    [0] => wp-example.ngrok.app
                )

        )

    [body:protected] => {"@context":["https://www.w3.org/ns/activitystreams",{"ostatus":"http://ostatus.org#","atomUri":"ostatus:atomUri","inReplyToAtomUri":"ostatus:inReplyToAtomUri","conversation":"ostatus:conversation","sensitive":"as:sensitive","toot":"http://joinmastodon.org/ns#","votersCount":"toot:votersCount"}],"id":"https://masto.example.app/users/mastodon/statuses/110082446069839861/activity","type":"Create","actor":"https://masto.example.app/users/mastodon","published":"2023-03-25T06:13:44Z","to":["https://www.w3.org/ns/activitystreams#Public"],"cc":["https://masto.example.app/users/mastodon/followers","https://wp-example.ngrok.app/author/wordpress/"],"object":{"id":"https://masto.example.app/users/mastodon/statuses/110082446069839861","type":"Note","summary":null,"inReplyTo":"https://wp-example.ngrok.app/sad-eros/","published":"2023-03-25T06:13:44Z","url":"https://masto.example.app/@mastodon/110082446069839861","attributedTo":"https://masto.example.app/users/mastodon","to":["https://www.w3.org/ns/activitystreams#Public"],"cc":["https://masto.example.app/users/mastodon/followers","https://wp-example.ngrok.app/author/wordpress/"],"sensitive":false,"atomUri":"https://masto.example.app/users/mastodon/statuses/110082446069839861","inReplyToAtomUri":"https://wp-example.ngrok.app/sad-eros/","conversation":"tag:masto.example.app,2023-03-18:objectId=35:objectType=Conversation","content":"\u003cp\u003e\u003cspan class=\"h-card\"\u003e\u003ca href=\"https://wp-example.ngrok.app/author/wordpress/\" class=\"u-url mention\"\u003e@\u003cspan\u003ewordpress\u003c/span\u003e\u003c/a\u003e\u003c/span\u003e $request\u003c/p\u003e","contentMap":{"en":"\u003cp\u003e\u003cspan class=\"h-card\"\u003e\u003ca href=\"https://wp-example.ngrok.app/author/wordpress/\" class=\"u-url mention\"\u003e@\u003cspan\u003ewordpress\u003c/span\u003e\u003c/a\u003e\u003c/span\u003e $request\u003c/p\u003e"},"attachment":[],"tag":[{"type":"Mention","href":"https://wp-example.ngrok.app/author/wordpress/","name":"@wordpress@wp-example.ngrok.app"}],"replies":{"id":"https://masto.example.app/users/mastodon/statuses/110082446069839861/replies","type":"Collection","first":{"type":"CollectionPage","next":"https://masto.example.app/users/mastodon/statuses/110082446069839861/replies?only_other_accounts=true\u0026page=true","partOf":"https://masto.example.app/users/mastodon/statuses/110082446069839861/replies","items":[]}}},"signature":{"type":"RsaSignature2017","creator":"https://masto.example.app/users/mastodon#main-key","created":"2023-03-25T06:13:44Z","signatureValue":"Ic9JGB+Q3DkYnx/LlGmOkW6eYJbfhqaQxaDEeL74urpOpKYwjFOfPSyHZkIiGW5kzqaAqC8OSp3Yzds441yoQZSLruWEpjL764o9J9FKkYM4cZr7lBasq8tdRtYLag0hEJtgt9yvXS7JiM3Tk4F/5XVrQ1B9D4OtQQwBtILB1zUifzocqi+x31VeV45fwkRI61yjFhhOE7NTHB0P9hfnrArP9OcCKYqISqOgH2qApGXVeZgs4hW5LMxjJyXG+KLBmyNFiaLBCyyAgiwKxgDzeOGMwbisEnSxCZwyLTuEeqAp7HUbj3FVQZUlbLVN/B678Q2jVakyN+mAe6sUHsKV1w=="}}
    [route:protected] => /activitypub/1.0/users/2/inbox
)

```

GET request for a `/convenience/` page (filtered out the irrelevant headers)
```
$_SERVER array
(
    [SERVER_SOFTWARE] => nginx/1.16.0
    [REQUEST_URI] => /convenience/
    [HTTP_X_ORIGINAL_HOST] => wp-example.ngrok.app
    [HTTP_SIGNATURE] => keyId=\"https://masto.example.app/actor#main-key\",algorithm=\"rsa-sha256\",headers=\"(request-target) host date accept\",signature=\"i8vqEeFPld/evOgJeKVW8X0zbmeJ76Mcyg3EAknbNREX+CP6CffkZ+jyVzhOnXQfWVgis1r0mC6c6aShLoxAkKo/Np9SbzRZptlay85tbEJ4Tx4jvC/CVoFmJ35x6zngic8+pXTaM33OGQRk9VO778xdqXrV1FQ7qQmddPymsx0uT3nbVTwPCsqyXp0f9bxnR5J8dzwj6mNzWFGUelqlgpFHytSU10ebY9i46iJ9OOBHl7McYJknZrdB42eabPcohUfRjvZjKbMvgJX+Ze/LdLYiEd0el0+ijb5veYzLp6VbGql7a+9VT0NYRCXUzFJlvLeEVvQHHrmGhJ1zRB8piQ==\"
    [HTTP_DATE] => Tue, 18 Apr 2023 02:31:19 GMT
    [HTTP_ACCEPT_ENCODING] => gzip
    [HTTP_ACCEPT] => application/activity+json, application/ld+json; profile=\"https://www.w3.org/ns/activitystreams\", text/html;q=0.1
    [HTTP_USER_AGENT] => http.rb/5.1.0 (Mastodon/4.0.2; +https://masto.example.app/)
    [CONTENT_LENGTH] => 
    [CONTENT_TYPE] => 
    [REQUEST_METHOD] => GET
    [REQUEST_TIME_FLOAT] => 1681785079.3728
    [REQUEST_TIME] => 1681785079
)
```

Reply:
Puh... I understand the idea behind that, but there are already three different versions of the `key` ID and (as far as I know) not every plattform does support it for `GET`.

I totally understand it for `POST` requests, and I am fine to sign also `GET` requests, but I am not really convinced by checking for signatures on receiving `GET` requests.

I fear that this might cause a lot of support questions!

Reply:
If only some GET requests are verified, the setting would not provide the intended security.

On a practical level, despite not having the actor url for these Get requests, they come from instance actors, and would be cached in transients after the first request.

Also, I haven't yet seen an implementation that didn't sign requests. It would have difficulty federating with Mastodon.

Reply:
But only for `POST` requests. The `GET` stuff is (as far as I can remember) an optional opt-in thing!?! At least I can query my Mastodon endpoints without signature headers.

And I am totally fine, sending the headers for all requests, I am only sceptical on checking signatures on incoming `GET` requests!

I am totally fine to add that feature with a feature flag, but at this state, I fear too much problems if it is a visible setting, everyone could opt-in.

Reply:
I also checked a lot of other platforms (calckey, miskey, friendi.ca), that all have no signature verification on `GET` requests (checked to load the profile URL and the follower list).

Reply:
I just checked the Mastodon Settings and couldn't find anything on `GET`-Request verification. Can you point me to any documentation about that?

I see that they add verifivation to all requests, but I could not find any documentation about verifying them.

Reply:
Just checked the Code... Mastodon seem to always use the KeyID to get and verify the signature, so at least that should work... still sceptical about the `GET` verification thing...

Reply:
> The GET stuff is (as far as I can remember) an optional opt-in thing

Exactly, this is the same model with ` \get_option( 'activitypub_use_secure_mode', '0' ) `

Reply:
Sorry if this wasn't clear: Verifying GET requests would ONLY happen if a site has activated Secure Mode!

Reply:
I understand that, but as I wrote earlier in the thread, I fear that a lot of people will activate it (because what could be wrong with being even more secure) and that we see a lot of issues on GitHub and in the WP.org forums, because AP is a complex topic and not every WP user understands the problems with shared hosting, content negotiation, signature verification and caching.

Reply:
That's why I vote for a const instead of a setting, that only people who know what they are doing activate it in the first place. If this first try works good we could re-add the setting afterwards. I would see it as a soft launch. Mastodon also has no setting for that and gives it in the responsibility of the host of a mastodon instance.

Reply:
And if not that way, than we need at least an opt-out... A const that disables the setting/feature when set.